### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by yujinjcho**
https://github.com/yujinjcho/rcruitme/issues/4

In this pr:
- lets you create a user on `/sign-up` using first, last, email, and pw
- can login with email and pw on `/sign-in`
- sign out with `/sign-out`

next todos:
- handling authenticated/unauthenticated requests better i.e. when you're logged in can't go to `/sign-in`, redirects
- add google/linkedin auth
- add email confirmation and user activated
- add user type (candidate, recruiter)
- not sure if user model fields need to be optionals / need to look into what info google/linkedin gives
- [implement persistence](https://github.com/yujinjcho/rcruitme/blob/4f053598f065dd27c234dca20ea7cebab4c809f7/app/modules/SilhouetteModule.scala#L58) for PasswordInfo, OathInfo etc

### CODE REVIEW

File: app/controllers/HomeController.scala [hunk 1]
```diff
@@ -1,17 +1,32 @@
...
-    Ok(views.html.index())
-  }
-  
-  def hello(name: String) = Action { implicit request: Request[AnyContent] =>
-    Ok(views.html.hello(name))
-  }
-}
+package controllers
+
+import scala.concurrent.Future
```
Comment (achiu8) [inline app/controllers/HomeController.scala]: Is there convention for grouping / ordering of imports?

File: app/controllers/HomeController.scala [hunk 2]
```diff
@@ -1,17 +1,32 @@
...
+
+import com.mohiva.play.silhouette.api.actions.SecuredRequest
+import com.mohiva.play.silhouette.api.LogoutEvent
+import com.mohiva.play.silhouette.api.Silhouette
+import javax.inject._
+import play.api.mvc._
+
+import utils.auth.DefaultEnv
+
+@Singleton
```
Comment (achiu8) [inline app/controllers/HomeController.scala]: Does this need to be singleton?

File: app/controllers/SignInController.scala [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import play.api.Configuration
+import play.api.i18n.{ I18nSupport, Messages }
+import play.api.mvc.{ AbstractController, AnyContent, ControllerComponents, Request }
+
+import forms.SignInForm
+import models.services.UserService
+import utils.auth.DefaultEnv
+
+class SignInController @Inject() (
+  components: ControllerComponents,
```
Comment (achiu8) [inline app/controllers/SignInController.scala]: Should we try to be consistent with repeated dependency param names (`cc` vs. `components`)?

File: app/controllers/SignInController.scala [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+      data => {
+        val credentials = Credentials(data.email, data.password)
+        credentialsProvider.authenticate(credentials).flatMap { loginInfo =>
+          val result = Redirect(routes.HomeController.index())
+          userService.retrieve(loginInfo).flatMap {
+            // TODO: handle non-activated users
+            case Some(user) =>
+              val c = configuration.underlying
+              silhouette.env.authenticatorService.create(loginInfo).map {
+                case authenticator if data.rememberMe =>
```
Comment (achiu8) [inline app/controllers/SignInController.scala]: Can this just be a conditional? Do we need cases?

```scala
if (!data.rememberMe) authenticator
else authenticator.copy(...)
```

File: app/controllers/SignUpController.scala [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  authInfoRepository: AuthInfoRepository,
+  authTokenService: AuthTokenService,
+  )(implicit ex: DatabaseExecutionContext)
+  extends AbstractController(components) with I18nSupport {
+
+  def view = silhouette.UnsecuredAction.async { implicit request: Request[AnyContent] =>
+    Future.successful(Ok(views.html.signUp(SignUpForm.form)))
+  }
+
+  def submit() = silhouette.UnsecuredAction.async { implicit request: Request[AnyContent] =>
```
Comment (achiu8) [inline app/controllers/SignUpController.scala]: What's the convention for parens on parameter-less methods that have side effects? `submit` on other controller doesn't have.

File: app/controllers/SignUpController.scala [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    SignUpForm.form.bindFromRequest.fold(
+      form => Future.successful(BadRequest(views.html.signUp(form))),
+      data => {
+        val redirect = Redirect(routes.SignUpController.view())
+        val loginInfo = LoginInfo(CredentialsProvider.ID, data.email)
+        userService.retrieve(loginInfo).flatMap {
+          case Some(user) =>
+            Future.successful(redirect.flashing("info" -> "Account exists already"))
+          case _ =>
```
Comment (achiu8) [inline app/controllers/SignUpController.scala]: Should we explicitly say `None`?

File: app/models/daos/UserDAO.scala [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        .on("userId" -> userID)
+        .as(userRowParser.single)
+    }
+  }(ec)
+
+  def save(user: User): Future[User] = Future {
+    val userId = db.withConnection { implicit conn =>
+      user match {
+        case User(_, Some(first), credentialId, Some(last), userType, Some(email)) =>
+          val query = s"""
```
Comment (achiu8) [inline app/models/daos/UserDAO.scala]: Should we try to be more consistent about defining intermediate values in these methods? Currently, `query` is variably: a `String`, a `SimpleSql`, not defined, called `result`.

File: app/models/User.scala [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+import com.mohiva.play.silhouette.api.{ Identity, LoginInfo }
+
+case class User(
+  userID: Int,
+  firstName: Option[String],
+  credentialId: String,
+  lastName: Option[String],
+  userType: String,
+  email: Option[String]
```
Comment (achiu8) [inline app/models/User.scala]: `first`, `last`, and `email` are currently not nullable in database. Are model fields supposed to match?

File: app/models/service/AuthTokenService.scala [hunk 9]
```diff
@@ -0,0 +1,33 @@
+package models.services
```
Comment (achiu8) [inline app/models/service/AuthTokenService.scala]: Should directory name match `services`?

File: conf/application.conf [hunk 10]
```diff
@@ -12,3 +12,25 @@ database.dispatcher {
...
+
+fixedConnectionPool = 9
+
+database.dispatcher {
+  executor = "thread-pool-executor"
+  throughput = 1
+  thread-pool-executor {
+    fixed-pool-size = ${fixedConnectionPool}
+  }
+}
```
Comment (achiu8) [inline conf/application.conf]: Looks like lines 19 to here are duplicated.

File: project/build.properties [hunk 11]
```diff
@@ -1 +1 @@
-sbt.version=1.2.3
+sbt.version=1.2.3
```
Comment (achiu8) [inline project/build.properties]: Any idea why this file is showing a diff?

File: app/controllers/HomeController.scala (hunk 1 above)
Comment (yujinjcho) [reply app/controllers/HomeController.scala]: I've been doing scala packages, java packages, everything else sorted in alphabetical order and then application packages. Don't remember where exactly I saw that. I'm fine standardizing on whatever.

File: app/controllers/HomeController.scala (hunk 2 above)
Comment (yujinjcho) [reply app/controllers/HomeController.scala]: I think should not be per https://github.com/google/guice/wiki/Scopes#choosing-a-scope

File: app/controllers/SignUpController.scala (hunk 5 above)
Comment (yujinjcho) [reply app/controllers/SignUpController.scala]: think should have parens

File: project/build.properties (hunk 11 above)
Comment (yujinjcho) [reply project/build.properties]: File I originally copied had weird newline diff that I updated
https://superuser.com/questions/1217622/how-can-i-remove-the-m-from-my-file-in-sublime-text-3

affects some other files too e.g. HomeController

**Comment by yujinjcho** [comment]
lemme know if you see anything else. I also updated the DAOs to use implicit ec (don't have to pass into Future) and added more type annotations.

**Pull request merged**